### PR TITLE
Custom logger

### PR DESCRIPTION
**Pull request by waahhhh**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | maybe
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

**first commit:**
I have sorted the code lines of the `GeneratorFactory` for the reason: put together what belongs together.

**second commit** (the primary reason)**:**
The code generator can accept a logger. See: https://docs.laminas.dev/laminas-di/codegen/#logging.
The default logger is `\Psr\Log\NullLogger`.  Currently you can't configure a logger.
I would like to get the ability to configure a logger.

### CODE REVIEW

File: docs/book/codegen.md [hunk 1]
```diff
@@ -75,6 +75,9 @@ following keys (unknown keys are ignored):
   this value is not provided, you will need to set it with the generator's
   `setOutputDirectory()` method before calling `generate()`.
 
+- `logger`: By default `\Psr\Log\NullLogger` is defined. See the
+  [Logging](#logging) section below for details.
```
Comment (froschdesign) [inline docs/book/codegen.md]: * The leading backslash for the classname can be omitted because it is only needed in namespaced code.
* The hint "below" should avoid because only in the current state of the document is the section below but nobody knows if that will be the case in the future. The link is enough. 

```suggestion
- `logger`: By default `Psr\Log\NullLogger` is defined. See the
  [Logging section](#logging) for details.
```

File: docs/book/codegen.md [hunk 2]
```diff
@@ -84,6 +87,7 @@ return [
             'aot' => [
                 'namespace' => 'AppAoT\Generated',
                 'directory' => __DIR__ . '/../gen',
+                'logger' => \Psr\Log\LoggerInterface::class,
```
Comment (froschdesign) [inline docs/book/codegen.md]: The same here: the leading backslash for the classname can be omitted because it is only needed in namespaced code.

File: docs/book/cookbook/aot-guide.md [hunk 3]
```diff
@@ -127,6 +127,7 @@ class ConfigProvider
                 'aot' => [
                     'namespace' => __NAMESPACE__ . '\\Generated',
                     'directory' => __DIR__ . '/../gen',
+                    'logger' => \Psr\Log\LoggerInterface::class,
```
Comment (froschdesign) [inline docs/book/cookbook/aot-guide.md]: This cookbook recipe does not use the logger therefore is the configuration unnecessary and misleading.

File: docs/book/cookbook/aot-guide.md [hunk 4]
```diff
@@ -330,6 +331,7 @@ class ConfigProvider
...
```
Comment (froschdesign) [inline docs/book/cookbook/aot-guide.md]: The same here.

**Comment by waahhhh** [comment]
The documentation has been improved. I also fixed the failing unit test that was broken on the branch develop.

File: test/GeneratedInjectorDelegatorTest.php [hunk 5]
```diff
@@ -24,7 +24,7 @@ public function testProvidedNamespaceIsNotAString()
             ->willReturn(true)
             ->shouldBeCalledTimes(1);
         $container->get('config')
-            ->willReturn(['dependencies' => ['auto' => ['aot' => ['namespace' => []]]]])
+            ->willReturn(['dependencies' => ['auto' => ['aot' => ['namespace' => 123]]]])
```
Comment (waahhhh) [inline test/GeneratedInjectorDelegatorTest.php]: Maybe I should revert this change and we should avoid to use `empty` in the `GeneratedInjectorDelegator` to check against an existing namespace.

File: src/CodeGenerator/InjectorGenerator.php [hunk 6]
```diff
@@ -185,6 +185,11 @@ public function getNamespace() : string
         return $this->namespace;
     }
 
+    public function getLogger() : LoggerInterface
```
Comment (tux-rampage) [inline src/CodeGenerator/InjectorGenerator.php]: This is exposing the logger unnecessarily. As far as I can see this is only used in the Unit test suite.
Therfore we should mark this method either `@internal` or remove it and test the logger assignment differently.

File: test/Container/GeneratorFactoryTest.php [hunk 7]
```diff
@@ -87,6 +89,33 @@ public function testSetsNamespaceFromConfig() : void
...
+                'auto' => [
+                    'aot' => [
+                        'logger' => 'MyCustomLogger',
+                    ],
+                ],
+            ],
+        ]);
+
+        $generator = (new GeneratorFactory())->create($container);
+        $this->assertInstanceOf(LoggerInterface::class, $generator->getLogger());
```
Comment (tux-rampage) [inline test/Container/GeneratorFactoryTest.php]: This is already covered by the Method's typehint and therefore not necessary

File: test/Container/GeneratorFactoryTest.php [hunk 8]
```diff
@@ -87,6 +89,33 @@ public function testSetsNamespaceFromConfig() : void
...
+        $this->assertNotInstanceOf(NullLogger::class, $generator->getLogger());
```
Comment (tux-rampage) [inline test/Container/GeneratorFactoryTest.php]: See the info above about the `getLogger()` implementation. Can we check this differently without exposing this dependency?

File: test/GeneratedInjectorDelegatorTest.php (hunk 5 above)
Comment (tux-rampage) [reply test/GeneratedInjectorDelegatorTest.php]: This was introduced on purpose since a Stringification of an array (that might come from a misconfiguration) would result in `"Array"` which would be a valid namespace name but possibly not be the user's intention. This test covers this case.

See: https://github.com/zendframework/zend-di/pull/51#discussion_r358204946

**Review by tux-rampage — CHANGES_REQUESTED** [review]
Hello @waahhhh 

thanks for your contribution :+1: 
Could you please check my remarks on some changes?

File: test/GeneratedInjectorDelegatorTest.php (hunk 5 above)
Comment (waahhhh) [reply test/GeneratedInjectorDelegatorTest.php]: I see the problem at this point.
We want to check if the content of `namespace` is a string.
But the problem is that in case of an empty array we already fall back to a default by validation.
I would suggest that we only set the default if `namespace` is not given and otherwise validate the content.

Comment (waahhhh) [reply test/GeneratedInjectorDelegatorTest.php]: maybe something like: 
````php
$namespace = $aotConfig['namespace'] ?? 'Laminas\Di\Generated';

if (! is_string($namespace)) {
    throw new InvalidServiceConfigException('Provided namespace is not a string.');
}
````

Comment (tux-rampage) [reply test/GeneratedInjectorDelegatorTest.php]: > We want to check if the content of namespace is a string.

Actually we want to check that a PHP array - even an empty one - (from a misgonfiguation) does not become `"Array"` **and** causes an exception, since this cannot be caught by static analysis at the moment.

Personally I'd prefer to convert the AOT config array into a strongly typed object instead of accessing the array directly so that static analysis and config loading catches them instead of the factory. But that exceeds the scope of this PR.

File: src/GeneratedInjectorDelegator.php [hunk 9]
```diff
@@ -22,7 +22,7 @@ public function __invoke(ContainerInterface $container, $name, callable $callbac
     {
         $config = $container->has('config') ? $container->get('config') : [];
         $aotConfig = $config['dependencies']['auto']['aot'] ?? [];
-        $namespace = empty($aotConfig['namespace']) ? 'Laminas\Di\Generated' : $aotConfig['namespace'];
+        $namespace = $aotConfig['namespace'] ?? 'Laminas\Di\Generated';
```
Comment (tux-rampage) [inline src/GeneratedInjectorDelegator.php]: Would you be OK to move this to a separate PR or is this required for this PR to pass?

**Review by tux-rampage — CHANGES_REQUESTED** [review]
@waahhhh Thanks for taking time to address this. Only one thing about the namespace config topic, otherwise it lgtm. :+1:

File: src/GeneratedInjectorDelegator.php (hunk 9 above)
Comment (waahhhh) [reply src/GeneratedInjectorDelegator.php]: For me the test was the only reason. I'll move it into a separate PR. :+1:
Please note that the test `GeneratedInjectorDelegatorTest::testProvidedNamespaceIsNotAString()` will still fail.

Comment (tux-rampage) [reply src/GeneratedInjectorDelegator.php]: > Please note that the test `GeneratedInjectorDelegatorTest::testProvidedNamespaceIsNotAString()` will still fail.

It's fine for me to handle this in a separate Issue/PR since it's not related to the configurable logger feature and it is not an issue introduced with this PR.

@froschdesign are there any objections with handling it this way?

**Review by tux-rampage — APPROVED** [review]

**Comment by tux-rampage** [comment]
Thank you very much @waahhhh 
I'm just checking back with the technical steering comitee before merging this.

File: docs/book/codegen.md [hunk 10]
```diff
@@ -75,6 +75,9 @@ following keys (unknown keys are ignored):
   this value is not provided, you will need to set it with the generator's
   `setOutputDirectory()` method before calling `generate()`.
 
+- `logger`: By default `Psr\Log\NullLogger` is defined. See the
```
Comment (michalbundyra) [inline docs/book/codegen.md]: I'd say "is used" instead of "is defined".

Also, I would be more descriptive here - `logger` must be resolvable in container and must be an instance of `Psr\Log\LoggerInterface`.

**Review by michalbundyra — COMMENTED** [review]
@waahhhh / @tux-rampage in general it looks good for me, just small comment about documentation.

File: docs/book/codegen.md (hunk 10 above)
Comment (tux-rampage) [reply docs/book/codegen.md]: Agree. 

@waahhhh could you please address this?

**Review by tux-rampage — APPROVED** [review]

**Comment by tux-rampage** [comment]
@michalbundyra lgtm, can you confirm that your feedback is properly adressed?

**Review by michalbundyra — APPROVED** [review]
🚢
